File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

import Data.List

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit = sum $ listOfMultiples factors limit

listOfMultiples :: [Integer] -> Integer -> [Integer]
listOfMultiples factors limit = unique allMultiples
  where allMultiples = concatMap (`multiples` limit) factors

multiples :: Integer -> Integer -> [Integer]
multiples factor limit = takeWhile (< limit) [factor,factor*2..]

unique :: (Ord a) => [a] -> [a]
unique = map head . group . sort

File: cpp/sum-of-multiples/sum_of_multiples.cpp
#include "sum_of_multiples.h"
#include <unordered_set>

int sum_of_multiples::to(const std::vector<int> factors, int max_num) {
    std::unordered_set<int> multiples;

    for (int factor : factors) {
        int multiple = factor;
        while (multiple < max_num) {
            multiples.insert(multiple);
            multiple += factor;
        }
    }

    int sum = 0;
    for (int multiple : multiples) {
        sum += multiple;
    }

    return sum;
}
